src/graphs/hopcroft_karp.rs

```diff
@@ -0,0 +1,93 @@
+//! # Hopcroft Karp algorithm for Bipartite Matching
+
+use std::collections::VecDeque;
+
+/// # Example
+/// ```
+/// use programming_team_code_rust::graphs::hopcroft_karp::hopcroft_karp;
+///
+/// let mut adj = vec![vec![]; 2];
+/// for (u, v) in [(0, 0), (0, 2), (1, 2)] {
+///    adj[u].push(v);
+/// }
+///
+/// let (matching_siz, l_to_r, r_to_l, mvc_l, mvc_r) = hopcroft_karp(&adj, 3);
+///
+/// assert_eq!(matching_siz, 2);
+/// assert_eq!(l_to_r, [Some(0), Some(2)]);
+/// assert_eq!(r_to_l, [Some(0), None, Some(1)]);
+/// assert_eq!(mvc_l, [true, true]);
+/// assert_eq!(mvc_r, [false, false, false]);
+/// ```
+///
+/// # Complexity
+/// - V: number of vertices; V = lsz + rsz
+/// - E: number of edges
+/// - Time: O(V + E * sqrt(v))
+/// - Space: O(V)
+#[allow(clippy::type_complexity)]
+pub fn hopcroft_karp(
```

Comment:
I like this name over "max matching" because it specifies which matching alg it is (there are multiple)